### ISSUE TITLE
CockroachDBDatabase doesn't report enum types on columns correctly anymore

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14597]
CockroachDB generated code no longer correctly links columns to their enum types. The types are now declared as follows:

```java
/**
 * @deprecated Unknown data type. If this is a qualified, user-defined type,
 * it may have been excluded from code generation. If this is a built-in
 * type, you can define an explicit {@link org.jooq.Binding} to specify how
 * this type should be handled. Deprecation can be turned off using
 * {@literal <deprecationOnUnknownTypes/>} in your code generator
 * configuration.
 */
@Deprecated
public final TableField<TBookRecord, Object> STATUS = createField(DSL.name("status"), org.jooq.impl.DefaultDataType.getDefaultDataType("\"postgres\".\"u_book_status\""), this, "");
```

They still work (including casts), because we still reference the type name in the column declaration, which is why our integration tests haven't picked this up yet. Obviously, the `Object` type is annoying for users, so users might be inclined to attach a `Converter` to it, e.g.

```
/**
 * The column <code>public.t_t.u</code>.
 */
public final TableField<TTRecord, E> U = createField(DSL.name("u"), org.jooq.impl.DefaultDataType.getDefaultDataType("\"postgres\".\"u_u\""), this, "", new EnumConverter<Object, E>(Object.class, E.class));
```

But now, the query is no longer generated correctly. The enum type cast is removed:

```
org.jooq.exception.DataAccessException: SQL [select "public"."t_t"."i", "public"."t_t"."u" from "public"."t_t" where "public"."t_t"."u" = ? -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT]; ERROR: cannot create enum of unspecified type
	at org.jooq_3.18.0-SNAPSHOT.POSTGRES_13.debug(Unknown Source) ~[?:?]
	at org.jooq.impl.Tools.translate(Tools.java:3423) ~[classes/:?]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:746) ~[classes/:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:371) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:290) ~[classes/:?]
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2839) ~[classes/:?]
	at org.jooq.testscripts.JDBC.main(JDBC.java:39) ~[test-classes/:?]
Caused by: org.postgresql.util.PSQLException: ERROR: cannot create enum of unspecified type
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190) ~[postgresql-42.5.3.jar:42.5.3]
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:177) ~[postgresql-42.5.3.jar:42.5.3]
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:219) ~[classes/:?]
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4708) ~[classes/:?]
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:236) ~[classes/:?]
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:357) ~[classes/:?]
	... 3 more
```

This seems to be a regression. To be investigated.

**Comment by lukaseder** [comment]
This is the query fetching the columns:

```sql
select
  information_schema.columns.column_name,
  information_schema.columns.ordinal_position,
  case
    when (
      information_schema.columns.crdb_sql_type like 'INTERVAL%YEAR%'
      or information_schema.columns.crdb_sql_type like 'INTERVAL%MONTH%'
    ) then 'INTERVAL YEAR TO MONTH'
    when (
      information_schema.columns.crdb_sql_type like 'INTERVAL%DAY%'
      or information_schema.columns.crdb_sql_type like 'INTERVAL%HOUR%'
      or information_schema.columns.crdb_sql_type like 'INTERVAL%MINUTE%'
      or information_schema.columns.crdb_sql_type like 'INTERVAL%SECOND%'
    ) then 'INTERVAL DAY TO SECOND'
    else information_schema.columns.crdb_sql_type
  end as data_type,
  coalesce(
    information_schema.columns.character_maximum_length,
    case
      when information_schema.columns.udt_name in (
        '_varchar', '_bpchar', '_char'
      ) then (pg_catalog.pg_attribute.atttypmod - 4)
    end
  ) as character_maximum_length,
  coalesce(
    coalesce(
      information_schema.columns.datetime_precision,
      information_schema.columns.numeric_precision,
      case
        when (
          information_schema.columns.udt_name in (
            '_time', '_timetz', '_timestamp', '_timestamptz'
          )
          and pg_catalog.pg_attribute.atttypmod >= 0
        ) then pg_catalog.pg_attribute.atttypmod
        when information_schema.columns.udt_name = '_numeric' then (((pg_catalog.pg_attribute.atttypmod - 4) >> 16) & 65535)
      end
    ),
    cast(nullif(
      regexp_replace(case
        when upper(information_schema.columns.crdb_sql_type) in (
          'TIMESTAMP', 'TIMESTAMP(TZ)'
        ) then (information_schema.columns.crdb_sql_type || '(6)')
        else information_schema.columns.crdb_sql_type
      end, '.*\((\d+)\).*', '\1'),
      information_schema.columns.crdb_sql_type
    ) as int4)
  ) as numeric_precision,
  coalesce(
    information_schema.columns.numeric_scale,
    case
      when information_schema.columns.udt_name = '_numeric' then ((pg_catalog.pg_attribute.atttypmod - 4) & 65535)
    end
  ) as numeric_scale,
  case
    when (
      information_schema.columns.is_identity = 'YES'
      or (
        count(*) filter (where information_schema.columns.is_identity = 'YES') over () = 0
        and lower(information_schema.columns.column_default) ilike 'nextval(''%''::%)'
      )
    ) then 'YES'
  end as is_identity,
  information_schema.columns.is_nullable,
  case
    when information_schema.columns.is_generated in (
      'YES', 'ALWAYS'
    ) then information_schema.columns.generation_expression
    else null
  end as generation_expression,
  pg_catalog.pg_attribute.attgenerated as attgenerated,
  case
    when (
      information_schema.columns.is_identity = 'YES'
      or (
        count(*) filter (where information_schema.columns.is_identity = 'YES') over () = 0
        and lower(information_schema.columns.column_default) ilike 'nextval(''%''::%)'
      )
    ) then null
    else information_schema.columns.column_default
  end as column_default,
  coalesce(
    information_schema.columns.domain_schema,
    substring(
      information_schema.columns.crdb_sql_type,
      1,
      nullif(
        (position('.' in information_schema.columns.crdb_sql_type) - 1),
        -1
      )
    )
  ) as udt_schema,
  coalesce(
    information_schema.columns.domain_name,
    case
      when information_schema.columns.data_type = 'ARRAY' then substring(pg_catalog.pg_type.typname, 2)
      else information_schema.columns.udt_name
    end
  ) as udt_name,
  pg_catalog.pg_description.description
from information_schema.columns
  join (
    pg_catalog.pg_attribute
      join (
        pg_catalog.pg_class as alias_100646588
          join pg_catalog.pg_namespace as alias_117804412
            on alias_100646588.relnamespace = alias_117804412.oid
      )
        on pg_catalog.pg_attribute.attrelid = alias_100646588.oid
    )
    on (
      pg_catalog.pg_attribute.attname = information_schema.columns.column_name
      and alias_100646588.relname = information_schema.columns.table_name
      and alias_117804412.nspname = information_schema.columns.table_schema
    )
  join pg_catalog.pg_type
    on pg_catalog.pg_attribute.atttypid = pg_catalog.pg_type.oid
  left outer join pg_catalog.pg_description
    on (
      pg_catalog.pg_description.objoid = pg_catalog.pg_attribute.attrelid
      and pg_catalog.pg_description.objsubid = information_schema.columns.ordinal_position
    )
where (
  information_schema.columns.table_schema = 'public'
  and information_schema.columns.table_name = 't_book'
  and not (cast(information_schema.columns.is_hidden as string) in (
    'yes', 'YES', 'TRUE', 'enabled', '1', '1.0', 't', 'T', 'true', 'y', 'Y', 'ENABLED', 
    'on', 'ON'
  ))
)
order by information_schema.columns.ordinal_position
```

There seems to be a mixup in the name of the `udt_schema` and `udt_catalog`:

![image](https://user-images.githubusercontent.com/734593/217594533-38dbb21d-3bc0-455a-bb24-89c9ccc44732.png)

The reason being that CockroachDB now qualifies `crdb_sql_type` with `catalog.schema.name`, not `schema.name`

**Comment by lukaseder** [comment]
We did this workaround because of https://github.com/cockroachdb/cockroach/issues/60145, which has now been fixed, so `udt_schema` is now correct.

**Comment by lukaseder** [comment]
Instead of using `substring()`, this regex should work to extract the `udt_schema`:

```sql
regexp_replace(crdb_sql_type, '(.*\.)?(.*?)\..*', '\2')
```

For backwards compatibility:

- We won't rely on `information_schema.columns.udt_schema` yet
- We will match any level of qualification in `crdb_sql_type`, including `catalog.schema.type` and `schema.type` (but also more, in case that ever happens. The penultimate identifier part is assumed to be the schema).

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.8 (https://github.com/jOOQ/jOOQ/issues/14603), and 3.16.14 (https://github.com/jOOQ/jOOQ/issues/14604).

Will release these versions today or tomorrow.
